src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,637 @@
+
+    /**
+     * @return static
+     */
+    abstract public function prepend(string ...$prefix): self;
+
+    /**
+     * @return static
+     */
+    public function repeat(int $multiplier): self
+    {
+        $str = clone $this;
+        $str->string = str_repeat($str->string, 0 > $multiplier ? 0 : $multiplier);
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function replace(string $from, string $to): self;
+
+    /**
+     * @param string|callable $to
+     *
+     * @return static
+     */
+    abstract public function replaceMatches(string $fromPattern, $to): self;
+
+    /**
+     * @return static
+     */
+    abstract public function slice(int $start = 0, int $length = null): self;
+
+    /**
+     * @return static
+     */
+    public function snake(): self
+    {
+        $str = $this->camel()->title();
+        $str->string = strtolower(preg_replace(['/([A-Z]+)([A-Z][a-z])/', '/([a-z\d])([A-Z])/'], '\1_\2', $str->string));
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function splice(string $replacement, int $start = 0, int $length = null): self;
+
+    /**
+     * @return static[]
+     */
+    public function split(string $delimiter, int $limit = null, int $flags = null): array
+    {
+        if (null === $flags) {
+            throw new \TypeError('Split behavior when $flags is null must be implemented by child classes.');
+        }
+
+        if ($this->ignoreCase) {
+            $delimiter .= 'i';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (false === $chunks = preg_split($delimiter, $this->string, $limit, $flags)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Splitting failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Splitting failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        $str = clone $this;
+
+        if (self::PREG_SPLIT_OFFSET_CAPTURE & $flags) {
+            foreach ($chunks as &$chunk) {
+                $str->string = $chunk[0];
+                $chunk[0] = clone $str;
+            }
+        } else {
+            foreach ($chunks as &$chunk) {
+                $str->string = $chunk;
+                $chunk = clone $str;
+            }
+        }
+
+        return $chunks;
+    }
+
+    /**
+     * @param string|string[] $prefix
+     */
+    public function startsWith($prefix): bool
+    {
+        foreach ((array) $prefix as $prefix) {
+            if ($this->startsWith((string) $prefix)) {
```

Comment:
This looks like infinite recursion.

Reply:
Looks like, but it's not: this method must be overridden in concrete child classes to handle the case where `is_string($prefix)`.

Reply:
I don't think so? The abstract implementation will just return false, and all is well?

Reply:
Only empty array will return `false`, any string will result infinite recursion.

Reply:
That's true. The cast to an array should be replaced by an `is_array` guard?

Reply:
there is nothing to do here, the code works as expected, all child classes implement this correctly

Reply:
> all child classes implement this correctly

All **your** child classes. But this is non-internal class - anyone can extend it and be surprised by this infinite recursion and implicit requirements of this class. This foreach could be extracted to separate protected method (or just used inline in descendants - there is not much logic here), so `startsWith()` could be abstract.

Reply:
There are a million ways to make a buggy child class. I don't think this is worth it.

Reply:
> There are a million ways to make a buggy child class.

Providing base class with infinite recursion bomb is definitely helpful. ;)

Reply:
My reasoning: extending this base class should be extremely rare, thus it should be optimized for speed instead of extensibility. People that still do extend it will figure out by looking at existing implementation at some point.

Reply:
If performance is the goal, I suggest to inline this foreach in all (2?) implementations in child classes. Pros:

1. Avoid redundant function call (micropotimization).
2. More readable and easier to understand code.
3. More clear API - method will be abstract, so it will be obvious that you need to implement it in child class.

IMO right now this class is overusing inheritance to create some twisted relation between child and parent class. I'm sure people will report this as bug, because this is confusing - simplifying this may help avoid unnecessary support requests.

Reply:
There are 3 child classes and 4 methods. I added a guard clause.